pkg/operator/ceph/csi/spec.go

```diff
@@ -470,14 +473,65 @@ func startDrivers(namespace string, clientset kubernetes.Interface, ver *version
 	}
 
 	if ver.Major > KubeMinMajor || (ver.Major == KubeMinMajor && ver.Minor >= provDeploymentSuppVersion) {
-		err = createCSIDriverInfo(clientset, RBDDriverName, ownerRef)
-		if err != nil {
-			return errors.Wrapf(err, "failed to create CSI driver object for %q", RBDDriverName)
+		if EnableRBD {
+			err = createCSIDriverInfo(clientset, RBDDriverName, ownerRef)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create CSI driver object for %q", RBDDriverName)
+			}
 		}
-		err = createCSIDriverInfo(clientset, CephFSDriverName, ownerRef)
-		if err != nil {
-			return errors.Wrapf(err, "failed to create CSI driver object for %q", CephFSDriverName)
+		if EnableCephFS {
+			err = createCSIDriverInfo(clientset, CephFSDriverName, ownerRef)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create CSI driver object for %q", CephFSDriverName)
+			}
+		}
+	}
+	return nil
+}
+
+func stopDrivers(namespace string, clientset kubernetes.Interface) error {
+	if !EnableRBD {
+		logger.Info("CSI Ceph RBD driver disabled ")
+		err := deleteCSIDriverResources(
+			namespace, clientset, "csi-rbdplugin", "csi-rbdplugin-provisioner", "csi-rbdplugin-metrics", RBDDriverName)
+		if err != nil && apierrors.IsNotFound(err) {
+			return nil
+		}
+		logger.Info("successfully removed CSI Ceph RBD driver")
+	}
+
+	if !EnableCephFS {
+		logger.Info("CSI CephFS driver disabled")
+		err := deleteCSIDriverResources(
+			namespace, clientset, "csi-cephfsplugin", "csi-cephfsplugin-provisioner", "csi-cephfsplugin-metrics", CephFSDriverName)
+		if err != nil && apierrors.IsNotFound(err) {
+			return nil
 		}
+		logger.Info("successfully removed CSI CephFS driver")
+	}
+	return nil
+}
+
+func deleteCSIDriverResources(
+	namespace string, clientset kubernetes.Interface, daemonset string, deployment string, service string, driverName string) error {
+
+	if err := k8sutil.DeleteDaemonset(clientset, namespace, daemonset); err != nil {
+		logger.Errorf("failed to delete the %s. %v", daemonset, err)
```

Comment:
Before printing an error, you need to check for the error IsNotFound()

Reply:
The check is actually backwards since it checks for `IsNotFound()` instead of `!IsNotFound()`. But actually I just noticed that the `k8sutil` methods for deleting the daemonsets, deployments, and services will already ignore the IsNotFound error. So you can simply remove all these checks for IsNotFound() where the k8sutil helpers are called.